plane/src/drone/docker/commands.rs

```diff
@@ -187,3 +195,139 @@ pub async fn run_container(
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{names::Name, types::ResourceLimits};
+    use bollard::container::LogsOptions;
+    use futures_util::TryStreamExt;
+    use serde_json;
+
+    use super::*;
+
+    /// starts and runs command in container
+    /// with given container config, returns stdout + stderr as string
+    /// handles pulling image and removes container after it's done.
+    async fn run_container_with_config(
+        config: bollard::container::Config<String>,
+    ) -> anyhow::Result<String> {
+        let docker = bollard::Docker::connect_with_local_defaults()?;
+        let image = config.image.clone().unwrap();
+        pull_image(&docker, &image, None, false).await?;
+        let backend_name = BackendName::new_random();
+        let container_id = ContainerId::from(format!("plane-test-{}", backend_name));
+        let create_container_options = bollard::container::CreateContainerOptions {
+            name: container_id.to_string(),
+            ..Default::default()
+        };
+
+        docker
+            .create_container(Some(create_container_options), config)
+            .await?;
+        docker
+            .start_container::<String>(&container_id.to_string(), None)
+            .await?;
+        let mut wait_container = docker.wait_container::<String>(&container_id.to_string(), None);
+
+        let mut output = String::new();
+        while let (Some(msg), Some(_)) = (
+            docker
+                .logs::<String>(
+                    &container_id.to_string(),
+                    Some(LogsOptions {
+                        follow: true,
+                        stdout: true,
+                        stderr: true,
+                        ..Default::default()
+                    }),
+                )
+                .try_next()
+                .await?,
+            wait_container.try_next().await?,
+        ) {
+            output = output + &msg.to_string();
+        }
+
+        docker
+            .remove_container(&container_id.to_string(), None)
+            .await?;
+
+        Ok(output)
+    }
+
+    #[tokio::test]
+    async fn test_resource_limits() -> anyhow::Result<()> {
+        let docker = bollard::Docker::connect_with_local_defaults()?;
+        let backend_name = BackendName::new_random();
+        let _container_id = ContainerId::from(format!("plane-test-{}", backend_name));
+        let mut executor_config = ExecutorConfig::from_image_with_defaults("debian:bookworm");
+
+        let resource_limits: ResourceLimits = serde_json::from_str(
```

Comment:
nit: I prefer `serde_json::from_value(json! {` because it allows the compiler to enforce that the syntax is well-formed JSON. Not a huge deal here since it's a test, but it's nice to use it everywhere for consistency.

Reply:
oh cool i didn't realise that macro tokenized like that. sweet.